Implement Sphinx documentation

ahnaf-tahmid-chowdhury opened the pull request:
## Description

This PR introduces a comprehensive documentation setup for PyDAGMC using Sphinx.

### Key changes:

- Configured Sphinx with Myst-parser for Markdown support and `sphinx-book-theme`.
- Added necessary documentation dependencies (sphinx, sphinx-book-theme, myst-parser, nbsphinx, etc.) to `pyproject.toml`.
- Created initial page structure:
    - User Guide
    - Tutorial
    - Methodology
    - API Reference (autodoc for `pydagmc.dagnav`)
    - Getting Help
- Updated `README.md` to reflect new documentation and add a link.
- Configured `conf.py` for theme, extensions, and autodoc settings.

> Live demo: https://ahnaf-tahmid-chowdhury.github.io/pydagmc/

ahnaf-tahmid-chowdhury committed:
add docs

ahnaf-tahmid-chowdhury committed:
add documentation link

ahnaf-tahmid-chowdhury committed:
sphinx quickstart

ahnaf-tahmid-chowdhury committed:
nbsphinx_execute always

ahnaf-tahmid-chowdhury committed:
add ipykernel

ahnaf-tahmid-chowdhury committed:
add security file

ahnaf-tahmid-chowdhury committed:
add dagmc logo

ahnaf-tahmid-chowdhury committed:
add code of conduct

ahnaf-tahmid-chowdhury committed:
contact Paul Wilson

ahnaf-tahmid-chowdhury committed:
add contributing

ahnaf-tahmid-chowdhury committed:
add security file

ahnaf-tahmid-chowdhury committed:
remove dev comments

ahnaf-tahmid-chowdhury committed:
add sphinxcontrib-mermaid

ahnaf-tahmid-chowdhury committed:
add welcome page

ahnaf-tahmid-chowdhury committed:
add user guide

ahnaf-tahmid-chowdhury committed:
add templates

ahnaf-tahmid-chowdhury committed:
include root level files to doc

ahnaf-tahmid-chowdhury committed:
add getting help

ahnaf-tahmid-chowdhury committed:
add api

ahnaf-tahmid-chowdhury committed:
add tutorial

ahnaf-tahmid-chowdhury committed:
Clean up downloaded and created files

ahnaf-tahmid-chowdhury committed:
add methodology

ahnaf-tahmid-chowdhury committed:
better format

ahnaf-tahmid-chowdhury committed:
add readthedocs

ahnaf-tahmid-chowdhury committed:
add docs workflow

ahnaf-tahmid-chowdhury committed:
add pandoc

ahnaf-tahmid-chowdhury committed:
conda use base

ahnaf-tahmid-chowdhury committed:
python -m pip install .[docs]

ahnaf-tahmid-chowdhury committed:
conda run -n base pip install .[docs]

ahnaf-tahmid-chowdhury committed:
add conda run

ahnaf-tahmid-chowdhury committed:
remove html folder

ahnaf-tahmid-chowdhury committed:
reduce title size

ahnaf-tahmid-chowdhury committed:
better markdown format

ahnaf-tahmid-chowdhury committed:
NukeBox -> PyDAGMC

ahnaf-tahmid-chowdhury committed:
remove extra .

pshriwise commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,54 @@
+        run: |
+          cd docs
+          conda run -n base sphinx-build -b html source build
+          touch build/.nojekyll # Important for GitHub Pages
```
Can we expand on why this is important?

ahnaf-tahmid-chowdhury replied:
For creating the doc site through github?

ahnaf-tahmid-chowdhury replied:
https://ahnaf-tahmid-chowdhury.github.io/pydagmc/

pshriwise replied:
I mean in the comment :)

pshriwise commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,43 @@
+
+## Reporting
+
+If you witness or experience behavior that violates this Code of Conduct, please contact [Paul Wilson](mailto:paul.wilson@wisc.edu). All reports will be kept confidential.
```
@gonuke are you willing to be the point of contact for this?

pshriwise commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,199 @@
+# Contributing
+
+First off, thank you for considering contributing to PyDAGMC! We welcome contributions from everyone. Your help is essential for keeping it great.
```
```suggestion
First off, thank you for considering contributing to PyDAGMC! We welcome contributions from everyone.
```

pshriwise commented on SECURITY.md:
```diff
@@ -0,0 +1,24 @@
+# Security Policy
```
What do we think are the security concerns with `pydagmc`? We should what they are here, but leave room for unexpected concerns too. Realistically, I don't foresee any security (e.g. personal data, etc.) concerns with this tool.

pshriwise replied:
Yeah let's pull it unless there's meaningful content or concerns to address.

pshriwise reviewed (changes requested):
Initial review here. Overall, I think this looks quite good. It's much appreciated @ahnaf-tahmid-chowdhury!

ahnaf-tahmid-chowdhury committed:
Update docs/source/methodology.md

Co-authored-by: Patrick Shriwise <pshriwise@gmail.com>

ahnaf-tahmid-chowdhury committed:
Update CONTRIBUTING.md

Co-authored-by: Patrick Shriwise <pshriwise@gmail.com>

ahnaf-tahmid-chowdhury committed:
Update docs/source/getting-help.md

Co-authored-by: Patrick Shriwise <pshriwise@gmail.com>

ahnaf-tahmid-chowdhury committed:
Update docs/source/index.md

Co-authored-by: Patrick Shriwise <pshriwise@gmail.com>

ahnaf-tahmid-chowdhury committed:
refactor: relocate data model sections to new file

The 'EntitySets', 'Essential MOAB Tags', and 'Implicit Relationships' sections were moved from the methodology page to a new data-model page to improve the documentation's structure.

A link to the new page has been added to the table of contents.

ahnaf-tahmid-chowdhury commented on docs/source/methodology.md:
```diff
@@ -0,0 +1,190 @@
+
+At the core, DAGMC geometry is organized using MOAB **Entity Sets**. These are collections of other entities, which can include mesh elements (like triangles) or other entity sets, forming a hierarchical structure. PyDAGMC maps these entity sets to its `Group`, `Volume`, and `Surface` classes.
+
+### Essential MOAB Tags Used by DAGMC
```
created new file called data-model and moved the contents there

gonuke replied:
I don't see that file in this PR?

ahnaf-tahmid-chowdhury committed:
Refactor tutorial and methodology documentation for clarity and structure

- Updated the title and introductory text in the tutorial to enhance engagement and clarity.
- Improved the organization of the tutorial sections, emphasizing key functionalities of PyDAGMC.
- Added a new section in the tutorial detailing the internal operations of PyDAGMC, providing insights into model loading, entity access, triangle data handling, material assignment, entity creation, and volume calculation.

ahnaf-tahmid-chowdhury commented on docs/source/methodology.md:
```diff
@@ -0,0 +1,190 @@
+  * Supports adding (`add_set`) and removing (`remove_set`) entities.
+  * The `merge()` method allows combining two groups with the same name.
+
+## Key Operations and Their Implementation
```
Moved to tutorials

ahnaf-tahmid-chowdhury committed:
chore: remove unused sections from issue and pull request templates

ahnaf-tahmid-chowdhury commented:
@pshriwise I think you are suggesting removing the PyDAGMC Class Structure section from the methodology and keeping DAGMC Data Model there. If that’s correct, please let me know and I’ll update it accordingly.

pshriwise commented on docs/source/methodology.md:
```diff
@@ -0,0 +1,100 @@
+    C ==> D{"CATEGORY_TAG Present?"}
+    D ==> E["Create Objects"]
+    E ==> F["Assign IDs and Tags"]
+    F ==> G["Build Implicit Relationships"]
+    G ==> H["Enable Queries"]
+    H ==> I["Extract Geometry"]
```
A lot of this is done as part of MOAB. I don't think it's important to break down a file load in this kind of detail here. Mainly what users will be interested in (in my opinion) is what they can do with a file once loaded. In short, an additional sentence above to capture all of this should suffice. something like:

"When a file is loaded, its embedded geometric topology relationships and metadata (material/property assignments) are loaded according to the conventions established by the DAGMC library."

ahnaf-tahmid-chowdhury committed:
remove security

ahnaf-tahmid-chowdhury committed:
update methodology

ahnaf-tahmid-chowdhury committed:
Merge remote-tracking branch 'upstream/main' into doc

gonuke commented on README.md:
```diff
@@ -6,80 +6,4 @@
 
 PyDAGMC is a Python interface for interacting with DAGMC .h5m files through the embedded topological relationships and metadata contained within. These interactions occur through a set of Python classes corresponding to DAGMC’s metadata Group (a grouping of volumes or surfaces), Volume, and Surface groupings in the mesh database. This interface is intended to provide a simple interface for obtain ...
 
-PyDAGMC classes provide the ability to perform basic queries as properties of the class instances. These queries include:
```
I think many users like to see some basic documentation like this in the README on GH so they don't have to dive into a complete documentation page.  I think it would be good to leave this here, but move the link to detailed documentation before this.

gonuke commented on docs/source/conf.py:
```diff
@@ -0,0 +1,132 @@
+
+project = 'PyDAGMC'
+copyright = '2023 - {}, The UW CNERG Team. All rights reserved'.format(datetime.date.today().year)
+author = 'Patrick Shriwise, Paul Wilson, et al'
```
There are other authors who deserve more recognition than me (e.g. Romano, Peterson).  We should either be comprehensive or find another way to list authors here.

gonuke commented on pyproject.toml:
```diff
@@ -30,18 +30,39 @@ classifiers = [
     "Operating System :: OS Independent",
     "Topic :: Scientific/Engineering",
     "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
```
Should we get rid of some of these old versions?

gonuke commented on docs/source/index.md:
```diff
@@ -0,0 +1,52 @@
+# Welcome to PyDAGMC
+
+PyDAGMC is a Python interface for interacting with DAGMC `.h5m` files
```
Can we link the word "DAGMC" to the DAGMC documentation (such as it is) here and throughout?

gonuke commented on docs/source/methodology.md:
```diff
@@ -0,0 +1,113 @@
+
+# Methodology
+
+PyDAGMC provides a Pythonic abstraction layer over PyMOAB for interacting with DAGMC (Direct Accelerated Geometry Monte Carlo) `.h5m` files. This section delves into the underlying concepts, data structures, and design choices that enable PyDAGMC's functionality.
```
I wonder if we should include a statement about what a DAGMC file actually is?  Something like:

```
A DAGMC file contains a surface mesh representation of a geometry, including 
metadata to define the topological relationships of that geometry and material
assignments, using MOAB's mesh file conventions.  The metadata conventions are 
specific to DAGMC.
```

We do get into it more in the details below, but perhaps some early statement like this is helpful context and provides some useful vocabulary and semantic nuance.

gonuke reviewed (changes requested):
I made a few suggestions here, but it looks like we're waiting on responses to other review comments still.

ahnaf-tahmid-chowdhury committed:
Update CONTRIBUTING.md

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

ahnaf-tahmid-chowdhury committed:
restore previous intro

ahnaf-tahmid-chowdhury committed:
short the title

ahnaf-tahmid-chowdhury committed:
update python

ahnaf-tahmid-chowdhury committed:
apply @gonuke's suggestions

ahnaf-tahmid-chowdhury committed:
Merge branch 'main' into doc

pshriwise commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,199 @@
+
+Before you begin, ensure you have the following installed:
+
+- **Python:** Version 3.8 or higher.
```
We've moved up to 3.9 after this was opened.
